sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -668,4 +672,31 @@ object DataSource extends Logging {
     }
     globPath
   }
+
+  /**
+   * Return all paths represented by the wildcard string.
+   * Follow [[InMemoryFileIndex]].bulkListLeafFile and reuse the conf.
+   */
+  private def getGlobbedPaths(
+      sparkSession: SparkSession,
+      fs: FileSystem,
+      hadoopConf: SerializableConfiguration,
+      qualified: Path): Seq[Path] = {
+    val paths = SparkHadoopUtil.get.expandGlobPath(fs, qualified)
+    if (paths.size <= sparkSession.sessionState.conf.parallelPartitionDiscoveryThreshold) {
+      SparkHadoopUtil.get.globPathIfNecessary(fs, qualified)
+    } else {
+      val parallelPartitionDiscoveryParallelism =
+        sparkSession.sessionState.conf.parallelPartitionDiscoveryParallelism
+      val numParallelism = Math.min(paths.size, parallelPartitionDiscoveryParallelism)
+      val expanded = sparkSession.sparkContext
```

Comment:
Why do this using a Spark job, instead of just a local thread pool?

I see this is the same thing done by `InMemoryFileIndex`, but it feels unnecessarily expensive.

Reply:
@vanzin Thanks for you reply.
```
Why do this using a Spark job, instead of just a local thread pool?
```
As the DFS generally deployed together with NodeManagers for better data locality, while using client mode and driver in different region with cluster, using a Spark job will resolve the problem of cross region interaction in our scenario.

Reply:
NMs are part of YARN, not HDFS.

This code will talk to HDFS's NameNodes; there is generally only one of those you'll be talking to. Latency here is not an issue, throughput is, so I still don't see why this shouldn't be done with a local thread pool.

Reply:
Yep, I means YARN and HDFS always deploy in same region, but driver we can't control because it's our customer's machine in client mode like spark sql or shell.
For example we deploy YARN and HDFS in Beijing CN, user use spark sql on Shanghai CN. 
Maybe this scenario shouldn't consider in this patch? What's your opinion @vanzin

Reply:
Where the driver is deployed is not of concern; first because, as I said, this is not about latency, but parallelizing multiple calls to the NN. Second, because if your driver is in a different network, it will have the same latency when talking to the executors as if it would talk directly to the NN, so you're not fixing anything, you're just adding an extra hop (= more latency).

Reply:
Thanks for your suggestion and detailed explanation, I'll reimplement this to local thread pool.

Reply:
Sorry for the late reply, finished in next commit.